pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,316 @@
+	csiDriverInformer := factory.Storage().V1().CSIDrivers()
+	csiDriverLister := csiDriverInformer.Lister()
+	volumeAttachmentInformer := factory.Storage().V1().VolumeAttachments()
+	volumeAttachmentLister := volumeAttachmentInformer.Lister()
+
+	plugMgr := &volume.VolumePluginMgr{}
+	fakeAttachDetachVolumeHost := volumetest.NewFakeAttachDetachVolumeHostWithCSINodeName(t,
+		tmpDir,
+		kubeClient,
+		fc.ProbeVolumePlugins(),
+		testNodeName,
+		csiDriverLister,
+		volumeAttachmentLister,
+	)
+
+	plugMgr.Host = fakeAttachDetachVolumeHost
+
+	pvLister := fakeframework.PersistentVolumeLister{
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: volumeName},
+			Spec: v1.PersistentVolumeSpec{
+				PersistentVolumeSource: v1.PersistentVolumeSource{
+					GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+						PDName:    "test-disk",
+						FSType:    "ext4",
+						Partition: 0,
+						ReadOnly:  false,
+					},
+				},
+				ClaimRef: &objectReference,
+			},
+		},
+	}
+
+	pvcLister := fakeframework.PersistentVolumeClaimLister{
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "test-pvc", Namespace: "default"},
+			Spec:       v1.PersistentVolumeClaimSpec{VolumeName: volumeName},
+			Status: v1.PersistentVolumeClaimStatus{
+				Phase: v1.ClaimBound,
+			},
+		},
+	}
+
+	tests := []struct {
+		desc                 string
+		podVolume            v1.Volume
+		pod                  *v1.Pod
+		enableCSIMigration   bool
+		wantVolume           v1.Volume
+		wantPersistentVolume v1.PersistentVolume
+		expectedErr          error
+	}{
+		{
+			desc:      "create volume spec happy case without csi migration enabled",
+			podVolume: volumeExampe,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+							},
+						},
+					},
+				},
+			},
+			enableCSIMigration: false,
+			wantVolume:         volumeExampe,
+		},
+		{
+			desc:      "create volume spec happy case with csi migration enabled but driver not installed",
+			podVolume: volumeExampeWithCSI,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "claim-name",
+									ReadOnly: false,
+								},
+							},
+						},
+					},
+				},
+			},
+			enableCSIMigration:   true,
+			wantPersistentVolume: persistentVolumeExampeWithCSI,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			actualSpec, _ := CreateVolumeSpec(test.podVolume, test.pod, nodeName, plugMgr, pvcLister, pvLister, intreeToCSITranslator, csiTranslator)
+			if test.enableCSIMigration {
```

Comment:
I think you could check on `wantVolume` or `wantPersistentVolume` being set instead of needing this extra parameter.

Reply:
You mean compare them to an empty struct by using cmp or reflect?

Reply:
I did this ``!reflect.DeepEqual(emptyVolume, test.wantPersistentVolume)``, feels weird to me but it reduce one param in the test struct.